examples/userapi/07_functions_on_subdomains.ipynb

```diff
@@ -0,0 +1,2999 @@
+    "\n",
+    "In many contexts, fields are required over only a subset of the computational `Grid`. In Devito, such fields can be implemented via the `Function`s on `SubDomain`s API, allocating arrays only over the required region of the `Grid`, thereby reducing memory consumption. This functionality is seamlessly compatible with MPI, as with any other Devito object, and these arrays are aligned with the domain decomposition under the hood.\n",
+    "\n",
+    "We will begin with some necessary imports."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "id": "18dea9f5-c610-4ac3-b76a-92a339fe55af",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import matplotlib.pyplot as plt\n",
+    "import matplotlib.animation as animation\n",
+    "import numpy as np\n",
+    "\n",
+    "from IPython.display import HTML\n",
+    "\n",
+    "from devito import SubDomain, Grid, Function, Eq, Operator, \\\n",
+    "    SparseFunction, ConditionalDimension, TimeFunction, solve, \\\n",
+    "    TensorTimeFunction, VectorTimeFunction, diag, grad, div\n",
+    "from examples.seismic import TimeAxis, RickerSource, Model, Receiver"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "55aa5f88-cd7b-4495-af2a-1112dfdffce1",
+   "metadata": {},
+   "source": [
+    "## Basic usage\n",
+    "\n",
+    "As usual, we define a subdomain template by subclassing `SubDomain` and overriding the `define` method. The `'middle'` subdomain we have defined in this case excludes two points from each side of the domain in each dimension."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "id": "04576474-fc3d-475e-9061-ef8f32fb04c9",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class Middle(SubDomain):\n",
+    "    name = 'middle'\n",
+    "\n",
+    "    def define(self, dimensions):\n",
+    "        return {d: ('middle', 2, 2) for d in dimensions}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f4ba4513-98ae-43fb-b1d8-35d9065632d0",
+   "metadata": {},
+   "source": [
+    "We then create a `Grid` and create an instance of `Middle`. We can then use this subdomain as the computational grid when defining a `Function`. For comparison, we also create a `Function` defined on a `Grid`."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "id": "72953a86-6714-4a50-8ea8-bd89f1d398cc",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "grid0 = Grid(shape=(11, 11), extent=(10., 10.))\n",
+    "\n",
+    "middle = Middle(grid=grid0)\n",
+    "\n",
+    "f = Function(name='f', grid=middle)  # Define Function on SubDomain\n",
+    "g = Function(name='g', grid=grid0)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "575aab51-f392-4d81-808a-5c1b41929ac5",
+   "metadata": {},
+   "source": [
+    "Inspecting the `shape` of `f`, it is apparent that it is smaller than the `Grid` on which `middle` is defined, instead sharing its shape with `middle`."
```

Comment:
fix english after "instead ..."

Reply:
Nominally grammatically correct, but have changed the wording to make it more readable